reactos/drivers/usb/usbohci_new/usbohci.c

```diff
@@ -0,0 +1,2728 @@
+
+    TransferedLen = 0;
+
+    do
+    {
+        TD->HwTD.gTD.Control.AsULONG = 0;
+        TD->HwTD.gTD.Control.DelayInterrupt = OHCI_TD_INTERRUPT_NONE;
+        TD->HwTD.gTD.Control.ConditionCode = OHCI_TD_CONDITION_NOT_ACCESSED;
+
+        if (TransferParameters->TransferFlags & USBD_TRANSFER_DIRECTION_IN)
+        {
+            TD->HwTD.gTD.Control.BufferRounding = 0;
+            TD->HwTD.gTD.Control.DirectionPID = OHCI_TD_DIRECTION_PID_IN;
+        }
+        else
+        {
+            TD->HwTD.gTD.Control.BufferRounding = TRUE;
+            TD->HwTD.gTD.Control.DirectionPID = OHCI_TD_DIRECTION_PID_OUT;
+        }
+
+        TD->HwTD.gTD.CurrentBuffer = 0;
+        TD->HwTD.gTD.NextTD = 0;
+        TD->HwTD.gTD.BufferEnd = 0;
+
+        RtlZeroMemory(&TD->HwTD.SetupPacket,
+                      sizeof(USB_DEFAULT_PIPE_SETUP_PACKET));
+
+        TD->HwTD.Padded[0] = 0;
+        TD->HwTD.Padded[1] = 0;
+
+        TD->Flags |= OHCI_HCD_TD_FLAG_PROCESSED;
+        TD->OhciTransfer = (ULONG)OhciTransfer;
+        TD->NextHcdTD = 0;
+
+        if (TransferParameters->TransferBufferLength)
+        {
+            MaxPacketSize = OhciEndpoint->EndpointProperties.TotalMaxPacketSize;
+
+            TransferedLen = OHCI_MapTransferToTD(OhciExtension,
+                                                 MaxPacketSize,
+                                                 TransferedLen,
+                                                 OhciTransfer,
+                                                 TD,
+                                                 SGList);
+        }
+        else
+        {
+            ASSERT(SGList->SgElementCount == 0);
+
+            TD->HwTD.gTD.CurrentBuffer = 0;
+            TD->HwTD.gTD.BufferEnd = 0;
+
+            TD->TransferLen = 0;
+        }
+
+        PrevTD = TD;
+
+        TD = OHCI_AllocateTD(OhciExtension, OhciEndpoint);
+        OhciTransfer->PendingTDs++;
+
+        PrevTD->HwTD.gTD.NextTD = TD->PhysicalAddress;
+        PrevTD->NextHcdTD = (ULONG)TD;
+    }
+    while (TransferedLen < TransferParameters->TransferBufferLength);
+
+    if (TransferParameters->TransferFlags & USBD_SHORT_TRANSFER_OK)
+    {
+        PrevTD->HwTD.gTD.Control.BufferRounding = TRUE;
+        OhciTransfer->Flags |= OHCI_TRANSFER_FLAGS_SHORT_TRANSFER_OK;
+    }
+
+    PrevTD->HwTD.gTD.Control.DelayInterrupt = OHCI_TD_INTERRUPT_IMMEDIATE;
+    PrevTD->HwTD.gTD.NextTD = TD->PhysicalAddress;
+    PrevTD->NextHcdTD = (ULONG)TD;
+
+    TD->HwTD.gTD.NextTD = 0;
+    TD->NextHcdTD = 0;
+
+    OhciTransfer->NextTD = TD;
+    OhciEndpoint->HcdTailP = TD;
+
+    OhciEndpoint->HcdED->HwED.TailPointer = TD->PhysicalAddress;
+
+    OHCI_EnableList(OhciExtension, OhciEndpoint);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+OHCI_SubmitTransfer(IN PVOID ohciExtension,
+                    IN PVOID ohciEndpoint,
+                    IN PVOID transferParameters,
+                    IN PVOID ohciTransfer,
+                    IN PVOID sgList)
```

Comment:
... and this PUSBPORT_SCATTER_GATHER_LIST?

Reply:
Changed.